File: ios/Classes/Layers/LayerContentHelper.swift
//
// Created by Mo on 2022/7/13.
//

import Foundation
import Flutter
import ArcGIS

class LayerContentHelper: NSObject {

    private static let instance: LayerContentHelper = LayerContentHelper()

    class func shareManager() -> LayerContentHelper {
        return instance;
    }

    private var secondaryLayers: Dictionary<String,
    [SecondaryLayer]> = Dictionary<String, [SecondaryLayer]>()

    func requireLayers(
            _ call: FlutterMethodCall,
            result: @escaping FlutterResult,
            layersController: LayersController
    ) -> Void {
        if let args = call.arguments as? Dictionary<String, Any> {
            if let layerId = args["layerId"] as? String {
                let foundLayer = layersController.getLayerByLayerId(layerId: layerId)
                if let fl = foundLayer as? AGSArcGISMapImageLayer {
                    if let foundLayerContents = secondaryLayers[layerId] {
                        sendLayersBack(result: result, layers: foundLayerContents)
                    } else {
                        loadSecondaryLayer(result: result, layerId: layerId, layer: fl)
                    }
                } else {
                    result(Dictionary<String, [SecondaryLayer]>())
                }
            } else {
                result(Dictionary<String, [SecondaryLayer]>())
            }
        } else {
            result(Dictionary<String, [SecondaryLayer]>())
        }
    }

    private func loadSecondaryLayer(result: @escaping FlutterResult, layerId: String, layer: AGSArcGISMapImageLayer) {
        var layerContents: [SecondaryLayer] = []
        for firstLayer in layer.subLayerContents {
            for secondLayer in firstLayer.subLayerContents {
                if (secondLayer is AGSArcGISMapImageSublayer) {
                    layerContents.append(SecondaryLayer(id: UUID().uuidString, layerContent: secondLayer))
                }
            }
        }
        secondaryLayers[layerId] = layerContents

        sendLayersBack(result: result, layers: layerContents)
    }

    private func sendLayersBack(result: @escaping FlutterResult, layers: [SecondaryLayer]) {
        var tmp: [Dictionary<String, Any>] = []
        layers.forEach { layer in
            tmp.append([
                "id": layer.id,
                "name": layer.layerContent.name,
                "visible": layer.layerContent.isVisible,
            ])
        }
        result(tmp)
    }

    func updateLayerVisibility(
            _ call: FlutterMethodCall,
            result: @escaping FlutterResult
    ) {
        if let args = call.arguments as? Dictionary<String, Any> {
            if let layerId = args["layerId"] as? String {
                let id = args["id"] as? String
                let foundLayer = secondaryLayers[layerId]
                let target = foundLayer?.first { layer in
                    return layer.id == id
                };
                if let visible = args["id"] as? Bool {
                    target?.layerContent.isVisible = visible
                    result(true)
                } else {
                    result(false)
                }
            } else {
                result(false)
            }
        } else {
            result(false)
        }

    }
}

private class SecondaryLayer {
    init(id: String, layerContent: AGSLayerContent) {
        self.id = id
        self.layerContent = layerContent
    }

    let id: String
    let layerContent: AGSLayerContent
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/layer_content/LayerContentHelper.kt
package com.valentingrigorean.arcgis_maps_flutter.layer_content

import com.esri.arcgisruntime.layers.ArcGISMapImageLayer
import com.esri.arcgisruntime.layers.ArcGISMapImageSublayer
import com.valentingrigorean.arcgis_maps_flutter.layers.LayersController
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.*

object LayerContentHelper {
    private val secondaryLayers: MutableMap<String, List<SecondaryLayer>> = mutableMapOf()

    fun requireLayers(
        call: MethodCall,
        result: MethodChannel.Result,
        layersController: LayersController
    ) {
        val layerId = call.argument<String?>("layerId")
        if (layerId == null) {
            result.success(emptyList<Map<String, Any>>())
        } else {
            val layer = layersController.getLayerByLayerId(layerId)
            if (layer is ArcGISMapImageLayer) {
                if (secondaryLayers.containsKey(layerId)) {
                    sendLayersBack(result, secondaryLayers[layerId].orEmpty())
                } else {
                    loadSecondaryLayer(result, layerId, layer)
                }
            } else {
                result.success(emptyList<Map<String, Any>>())
            }
        }
    }

    private fun loadSecondaryLayer(
        result: MethodChannel.Result,
        layerId: String,
        layer: ArcGISMapImageLayer
    ) {
        val layerContents: MutableList<SecondaryLayer> = mutableListOf()
        layer.subLayerContents.forEach { firstLayer ->
            firstLayer.subLayerContents.forEach { secondaryLayer ->
                if (secondaryLayer is ArcGISMapImageSublayer) {
                    layerContents.add(
                        SecondaryLayer(
                            id = UUID.randomUUID().toString(),
                            layerContent = secondaryLayer
                        )
                    )
                }
            }
        }
        secondaryLayers[layerId] = layerContents
        sendLayersBack(result, layerContents)
    }

    private fun sendLayersBack(
        result: MethodChannel.Result,
        layers: List<SecondaryLayer>,
    ) {
        val tmp: MutableList<Map<String, Any>> = mutableListOf()
        layers.forEach { item ->
            tmp.add(
                mapOf<String, Any>(
                    "id" to item.id,
                    "name" to item.layerContent.name,
                    "visible" to item.layerContent.isVisible
                )
            )
        }
        result.success(tmp)
    }

    fun updateLayerVisibility(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val layerId = call.argument<String?>("layerId")
        val id = call.argument<String?>("id")
        val foundLayer = secondaryLayers[layerId]
        if (foundLayer == null) {
            result.success(false)
        } else {
            val visible = call.argument<Boolean?>("visible")
            if (visible != null) {
                foundLayer.firstOrNull {
                    it.id == id
                }?.let {
                    it.layerContent.isVisible = visible
                }
            }

            result.success(true)
        }

    }
}
